src/widget/weffectparameterknobcomposed.cpp

```diff
@@ -1,37 +1,54 @@
 #include "widget/effectwidgetutils.h"
 #include "widget/weffectparameterknobcomposed.h"
 
-void WEffectParameterKnobComposed::setup(const QDomNode& node,
-                                         const SkinContext& context) {
-    WKnobComposed::setup(node, context);
-
-    EffectRackPointer pRack = EffectWidgetUtils::getEffectRackFromNode(
-            node, context, m_pEffectsManager);
-    EffectChainSlotPointer pChainSlot = EffectWidgetUtils::getEffectChainSlotFromNode(
-            node, context, pRack);
-    EffectSlotPointer pEffectSlot = EffectWidgetUtils::getEffectSlotFromNode(
-            node, context, pChainSlot);
-    EffectParameterSlotBasePointer pParameterSlot =
-            EffectWidgetUtils::getParameterSlotFromNode(
-                    node, context, pEffectSlot);
-    if (pParameterSlot) {
-        setEffectParameterSlot(pParameterSlot);
-    } else {
-        SKIN_WARNING(node, context)
-                << "EffectParameterKnobComposed node could not attach to effect parameter";
+namespace {
+const QString effectGroupSeparator = "_";
+const QString groupClose = "]";
+} // anonymous namespace
+
+void WEffectParameterKnobComposed::setupEffectParameterSlot(const ConfigKey& configKey) {
```

Comment:
IMO this function is quite hacky, but if you prefer doing it this way I don't care much either way.